flex-checks/src/main/java/org/sonar/flex/checks/ConditionalStructure.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * SonarQube Flex Plugin
+ * Copyright (C) 2010-2019 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.flex.checks;
+
+import com.sonar.sslr.api.AstNode;
+import com.sonar.sslr.api.Token;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import org.sonar.flex.FlexGrammar;
+import org.sonar.flex.FlexKeyword;
+
+import static java.util.Collections.singletonList;
+
+class ConditionalStructure {
+  static class BranchAndContent {
+    AstNode branch;
+    List<AstNode> content;
+    boolean onLiner;
+    BranchAndContent(AstNode branch, List<AstNode> content, boolean oneliner) {
+      this.branch = branch;
+      this.content = content;
+      this.onLiner = oneliner;
+    }
+  }
+  static BranchAndContent branchAndContentIf(AstNode branch, AstNode subStatement) {
+    return new BranchAndContent(branch, singletonList(subStatement), isOnelinerSubStatement(subStatement));
+  }
+  static BranchAndContent branchAndContentSwitch(AstNode branch, List<AstNode> directives) {
+    return new BranchAndContent(branch, directives, isOnelinerDirectives(directives));
+  }
+
+  final List<BranchAndContent> branches;
+  final boolean allBranchesArePresent;
+  final AstNode node;
+
+  ConditionalStructure(AstNode node, List<BranchAndContent> branches, boolean allBranchesArePresent) {
+    this.node = node;
+    this.branches = branches;
+    this.allBranchesArePresent = allBranchesArePresent;
+  }
+
+  boolean areAllEquivalentBranches() {
+    BranchAndContent first = branches.get(0);
+    return branches.stream().skip(1).allMatch(next -> SyntacticEquivalence.areEquivalent(first.content, next.content));
+  }
+
+  @FunctionalInterface
+  interface DuplicatedBranchCallback extends BiConsumer<AstNode,AstNode> {
+    void accept(AstNode branchFirstNode1, AstNode branchFirstNode2);
+  }
+
+  void forEachBranchDuplication(DuplicatedBranchCallback callback) {
+    for (BranchAndContent branch1 : branches) {
+      if (!allBranchesArePresent || !branch1.onLiner) {
```

Comment:
I believe that doesn't match the RSPEC and that it should be something like:
```
    boolean areAllEquivalent = areAllEquivalentBranches();
    if (allBranchesArePresent && areAllEquivalent) {
      return;
    }
    for (BranchAndContent branch1 : branches) {
      if (areAllEquivalent || !branch1.onLiner) {
```

Reply:
I do not understand what areAllEquivalent has to do with this.
And the test (!allBranchesArePresent || !branch1.onLiner) is the same as !(allBranchesArePresent && branch1.onLiner) which is exactly the exception described in the  RSPEC, no? 
I can change the test to the latter formulation in the code if you prefer it?

Reply:
Actually I prefer it. I changee it.